src/engine/map.cxx

```diff
@@ -361,10 +372,93 @@ void Map::highlightNode(const Point &isoCoordinates)
   {
     int index = isoCoordinates.x * _columns + isoCoordinates.y;
 
-    if (index >= 0 && index < _mapNodes.size())
+    if (index >= 0 && index < mapNodes.size())
     {
-      highlitNode = _mapNodes.at(index);
+      highlitNode = mapNodes[index].get();
       highlitNode->getSprite()->highlight(true);
     }
   }
 }
+
+void Map::saveMapToFile(const std::string &fileName)
+{
+  json j = json{{"columns", this->_columns}, {"rows", this->_rows}, {"mapNode", mapNodes}};
+
+  std::ofstream file(fileName);
+  file << j;
+}
+
+Map *Map::loadMapFromFile(const std::string &fileName)
+{
+  std::ifstream file(fileName);
+
+  if (file.fail())
+  {
+    LOG(LOG_ERROR) << "Could not load savegame file " << fileName;
+    return nullptr;
+  }
+
+  // check if json file can be parsed
+  json saveGameJSON = json::parse(file, nullptr, false);
+  if (saveGameJSON.is_discarded())
+  {
+    LOG(LOG_ERROR) << "Could not parse savegame file " << fileName;
+    return nullptr;
+  }
+
+  int columns = saveGameJSON.value("columns", -1);
+  int rows = saveGameJSON.value("rows", -1);
+
+  if (columns == -1 || rows == -1)
+    return nullptr;
+
+  Map *map = new Map(columns, rows);
+  map->initMap();
+
+  // set coordinates first
+  for (const auto &it : saveGameJSON["mapNode"].items())
+  {
+    Point coordinates = json(it.value())["coordinates"].get<Point>();
+    map->mapNodes[coordinates.x * columns + coordinates.y]->setCoordinates(coordinates);
+  }
+
+  // update all nodes to reflect new height differences
+  map->updateAllNodes();
+
+  // set tileIDs from savegame
+  for (const auto &it : saveGameJSON["mapNode"].items())
+  {
+    Point coordinates = json(it.value())["coordinates"].get<Point>();
+    map->mapNodes[coordinates.x * columns + coordinates.y]->setTileID(json(it.value())["tileID"].get<std::string>());
+  }
+
+  // update all nodes to have correct tiling for new tileIDs
+  map->updateAllNodes();
+
+  return map;
+}
+
+// JSON serializer for Point class
+void to_json(json &j, const Point &point) { j = json{{"x", point.x}, {"y", point.y}, {"z", point.z}, {"height", point.height}}; }
+
+// JSON serializer for MapNode class
+void to_json(json &j, const std::unique_ptr<MapNode> &m)
+{
+  if (m.get())
+  {
+    j = json{{"tileID", m->getTileID()}, {"coordinates", m->getCoordinates()}};
+  }
+  else
+  {
+    j = nullptr;
+  }
+}
```

Comment:
Add `file.close();`

Reply:
The C++ documentation for fstream says, that it's automatically close in it's destructor when it goes out of scope:
http://www.cplusplus.com/reference/fstream/fstream/close/

> Note that any open file is automatically closed when the fstream object is destroyed.

But i'll add it anyway... better be safe than sorry :)